modules/src/ics07_tendermint/client_def.rs

```diff
@@ -25,17 +36,117 @@ impl ClientDef for TendermintClient {
+        // check if a consensus state is already installed; if so it should
+        // match the untrusted header.
+
+        if let Some(cs) = ctx.consensus_state(&client_id, header.height()) {
+            //could the header height be zero ?
+            let consensus_state = downcast!(
+                cs => AnyConsensusState::Tendermint
+            )
+            .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?;
+
+            if consensus_state != ConsensusState::from(header.clone()) {
+                //freeze the client and return the installed consensus state
+                return Ok((
+                    client_state.with_set_frozen(header.height()),
+                    consensus_state,
+                ));
+            } else {
+                return Ok((client_state, consensus_state));
+            }
+        };
+
+        let latest_consensus_state =
+            match ctx.consensus_state(&client_id, client_state.latest_height) {
+                //could the header height be zero ?
+                Some(cs) => downcast!(
+                    cs => AnyConsensusState::Tendermint
+                )
+                .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?,
+                None => {
+                    return Err(Ics02Error::consensus_state_not_found(
+                        client_id.clone(),
+                        client_state.latest_height,
+                    ));
+                }
+            };
+
+        monotonicity_checks(latest_consensus_state, header.clone(), client_state.clone())
+            .map_err(Ics02Error::tendermint_handler_error)?;
 
-        // TODO: Additional verifications should be implemented here.
+        // check that the versions of the client state and the header match
+        if client_state.latest_height.revision_number != header.height().revision_number {
+            return Err(Error::mismatched_revisions(
+                client_state.latest_height.revision_number,
+                header.height().revision_number,
+            )
+            .into());
+        };
+
+        let trusted_consensus_state = match ctx.consensus_state(&client_id, header.trusted_height) {
+            Some(ts) => downcast!(
+                ts => AnyConsensusState::Tendermint
+            )
+            .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?,
+            None => {
+                return Err(Ics02Error::consensus_state_not_found(
+                    client_id,
+                    header.trusted_height,
+                ))
+            }
+        };
+
+        if header.height() == header.trusted_height.increment() {
+            //adjacent
+
+            // check that the header's trusted validator set is
+            // the next_validator_set of the trusted consensus state
+            if Set::hash(&header.validator_set) != trusted_consensus_state.next_validators_hash {
```

Comment:
This check should be done for both adjacent and non-adjacent. And you need the trusted validator instead:
```suggestion
            if Set::hash(&header.trusted_validator_set) != trusted_consensus_state.next_validators_hash {
```
This check should be done in the tendermint light client API.